### PR TITLE
JDBC sources: consolidate date/time values mapping

### LINKED ISSUE #17068: JDBC Sources: Consolidate putDate behavior, and potentially other methods
Multiple sources override the default putDate implementation (https://github.com/airbytehq/airbyte/blob/00927128459cf5d77aea9bf375163e347974d994/airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java#L126-L128), because it's incorrect. Specifically, it produces a full ISO8601 timestamp, when dates should just be `YYYY-MM-DD`.

Postgres and Mysql currently have the exact same implementation (https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java#L214-L217), and we intend to add that to Snowflake as well. Would be good to just modify the base class and fix this for all JDBC sources.

---

We should probably also check for other methods in a similar situation, and consolidate them too.

### PR DESCRIPTION
**Pull request by VitaliiMaltsev**
## What
closes [17068](https://app.zenhub.com/workspaces/db--dw-source-connectors-6333360e0a41155061efbcbd/issues/gh/airbytehq/airbyte/17068)
 Multiple JDBC sources produce a full ISO8601 timestamp during mapping of DATE or TIME datatypes, when date should just be `YYYY-MM-DD` and time should be `HH:mm:ss.SSSSSS`

Exception is OracleDB 
DATE in OracleDB could contain time also

## How
Updated mapping in JDBCSourceOperations

Note: PreparedStatement of Redshift, Snowflake and DB2 can not handle LocalDate/LocalTime java classes. Implementation for them is overriden in SourceOperations

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-db2/src/main/java/io.airbyte.integrations.source.db2/Db2SourceOperations.java
Comment (edgao) [inline airbyte-integrations/connectors/source-db2/src/main/java/io.airbyte.integrations.source.db2/Db2SourceOperations.java:96]: it looks like this code is identical to redshift / snowflake? Assuming I didn't just miss the difference - does it make sense to extract it to a shared method (maybe in DateTimeConverter?)

File: airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlJdbcSourceAcceptanceTest.java
Comment (edgao) [inline airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlJdbcSourceAcceptanceTest.java:24]: nitpick: this can be deleted

File: airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlSourceOperations.java
Comment (edgao) [inline airbyte-integrations/connectors/source-mysql/src/main/java/io/airbyte/integrations/source/mysql/MySqlSourceOperations.java:216]: ❤️ fewer method overrides!

File: airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSourceOperations.java
Comment (edgao) [inline airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSourceOperations.java:20]: let's add a comment explaining why this needs to be here
```suggestion
  /**
   * Oracle's DATE type can actually have time values, so we return a full timestamp.
   */
  protected void putDate(ObjectNode node, String columnName, ResultSet resultSet, int index) throws SQLException {
```

**Review by edgao — COMMENTED** [review]
couple comments - the questions around time/timetz handling are the big ones, other stuff is pretty minor.

File: airbyte-db/db-lib/src/test/java/io/airbyte/db/jdbc/TestJdbcUtils.java
Comment (VitaliiMaltsev) [reply airbyte-db/db-lib/src/test/java/io/airbyte/db/jdbc/TestJdbcUtils.java]: > I think both of these TODO comments are now done?

removed

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java
Comment (VitaliiMaltsev) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java:152]: > does this handle time with timezone values correctly? E.g. `01:23:45.67Z` (asking because it's using a LocalTime object)
> 
> (or am I misunderstanding how this method is supposed to be used?)

Nope, it's related to `TIME` datatype only and not to`TIMETZ`
I'm wondering that we don't have common mapping in JdbcSourceOperations for TIMETZ and TIMESTAMPTZ datatypes to use it as a cursor field. 
https://github.com/airbytehq/airbyte/blob/master/airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/JdbcSourceOperations.java#L66-L90
Currently it's implemented only for Postgres Source

I believe that we should have a separate task to use TIMETZ and TIMESTAMPTZ as a cursor for all JDBC sources

File: airbyte-integrations/connectors/source-cockroachdb/src/test-integration/java/io/airbyte/integrations/source/cockroachdb/CockroachDbSourceDatatypeTest.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-cockroachdb/src/test-integration/java/io/airbyte/integrations/source/cockroachdb/CockroachDbSourceDatatypeTest.java:330]: > I think we should keep the test cases for both time and timetz, since those are expected to behave differently? (i.e. eventually, we'll want to `discover` them as `TimeWithTimezone` and `TimeWithoutTimezone` respectively)

rolled it back

File: airbyte-integrations/connectors/source-db2/src/main/java/io.airbyte.integrations.source.db2/Db2SourceOperations.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-db2/src/main/java/io.airbyte.integrations.source.db2/Db2SourceOperations.java]: > it looks like this code is identical to redshift / snowflake? Assuming I didn't just miss the difference - does it make sense to extract it to a shared method (maybe in DateTimeConverter?)

extracted to the separate methods _putJavaSQLDate_() and _putJavaSQLTime_()

File: airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlJdbcSourceAcceptanceTest.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-mssql/src/test/java/io/airbyte/integrations/source/mssql/MssqlJdbcSourceAcceptanceTest.java]: > nitpick: this can be deleted

done

File: airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSourceOperations.java
Comment (VitaliiMaltsev) [reply airbyte-integrations/connectors/source-oracle/src/main/java/io/airbyte/integrations/source/oracle/OracleSourceOperations.java:23]: > let's add a comment explaining why this needs to be here

done

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java
Comment (edgao) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java:152]: got it. Does that mean that even after this PR is merged, `timetz` values (for postgres + cockroachdb) will still be `read` as full iso8601 timestamps?

**Review by edgao — APPROVED** [review]
lgtm assuming I understood correctly in https://github.com/airbytehq/airbyte/pull/20346#discussion_r1050020021

DB sources folks, any other comments?

**Comment by akashkulk** [comment]
I'm taking a quick look

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java
Comment (akashkulk) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java:152]: So this is fixing the `setTime` in the base class? If so, can you do the same and delete the redundant methods from inherited classes?

**Review by akashkulk — APPROVED** [review]
Looks good in general - just had some quick questions on the thread with Ed

File: airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java
Comment (VitaliiMaltsev) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java:152]: > got it. Does that mean that even after this PR is merged, `timetz` values (for postgres + cockroachdb) will still be `read` as full iso8601 timestamps?
> 
> edit: or is this the special handling you added in `putCockroachSpecialDataType`, and postgres was already behaving correctly?

yes, that's right

Comment (VitaliiMaltsev) [reply airbyte-db/db-lib/src/main/java/io/airbyte/db/jdbc/AbstractJdbcCompatibleSourceOperations.java:152]: > So this is fixing the `setTime` in the base class? If so, can you do the same and delete the redundant methods from inherited classes?

removed redundant methods from inherited classes for Postgres and MySQL

**Comment by VitaliiMaltsev** [comment]
/publish connector=connectors/source-clickhouse-strict-encrypt

**Comment by VitaliiMaltsev** [comment]
/publish connector=connectors/source-cockroachdb-strict-encrypt

**Comment by VitaliiMaltsev** [comment]
/publish connector=connectors/source-cockroachdb-strict-encrypt

**Comment by VitaliiMaltsev** [comment]
/publish connector=connectors/source-clickhouse-strict-encrypt

**Pull request merged**
